src/Nethermind/Nethermind.Core/Crypto/KeccakCache.cs

```diff
@@ -0,0 +1,226 @@
+                {
+                    // The value has been changed between reading and taking a lock.
+                    // Release the lock and compute.
+                    Volatile.Write(ref e.LockAndLength, lockAndLength);
+                    goto Compute;
+                }
+
+                // Take local copy of the payload and hash, to release the lock as soon as possible and make a key comparison without holding it.
+                // Local copy of 8+16+64 payload bytes.
+                Payload copy = e.Value;
+                // Copy Keccak256 directly to the local return variable, since we will overwrite if no match anyway.
+                keccak256 = e.Keccak256;
+
+                // Release the lock, by setting back to unnegated length.
+                Volatile.Write(ref e.LockAndLength, lockAndLength);
+
+                // Lengths are equal, the input length can be used without any additional operation.
+                if (input.Length == 32)
+                {
+                    // Hashing UInt256 or Hash256 which is Vector256
+                    if (Unsafe.As<byte, Vector256<byte>>(ref copy.Start) ==
+                        Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(input)))
+                    {
+                        // Current keccak256 is correct hash.
+                        return;
+                    }
+                }
+                else if (input.Length == 20)
+                {
+                    // Hashing Address
+                    ref byte bytes0 = ref copy.Start;
+                    ref byte bytes1 = ref MemoryMarshal.GetReference(input);
+                    // 20 bytes which is Vector128+uint
+                    if (Unsafe.As<byte, Vector128<byte>>(ref bytes0) == Unsafe.As<byte, Vector128<byte>>(ref bytes1) &&
+                        Unsafe.As<byte, uint>(ref Unsafe.Add(ref bytes0, Vector128<byte>.Count)) ==
+                            Unsafe.As<byte, uint>(ref Unsafe.Add(ref bytes1, Vector128<byte>.Count)))
+                    {
+                        // Current keccak256 is correct hash.
+                        return;
+                    }
+                }
+                else if (MemoryMarshal.CreateReadOnlySpan(ref copy.Start, input.Length).SequenceEqual(input))
+                {
+                    // Non 32 byte or 20 byte input; call SequenceEqual.
+                    // Current keccak256 is correct hash.
+                    return;
+                }
+            }
+        }
+
+    Compute:
+        keccak256 = ValueKeccak.Compute(input);
+
+        // Try lock and memoize
+        int length = Volatile.Read(ref e.LockAndLength);
+        // Negative value means that the entry is locked, set to int.MinValue to avoid confusion empty entry,
+        // since we are overwriting it anyway e.g. -0 would not be a reliable locked state.
+        if (length >= 0 && Interlocked.CompareExchange(ref e.LockAndLength, int.MinValue, length) == length)
```

Comment:
This is different from the original, that never tested whether it's locked or not, and always try to claim it. The current version not only changes this, but also, will take a lock ONLY if a potential update between 128 and 131 didn't change the length. This could be a good thing, meaning, that if an entry is updated hard with different lengths, this might save one update over a frequently updated entry.